### ISSUE TITLE
Upstreaming: Additional configuration for substitution

### ISSUE THREAD

**Issue by cpovirk** [jspecify/checker-framework#4]
https://github.com/jspecify/checker-framework/commit/d1aba1c548b757190b11d9c5a853c1ced0661a52

CF already lets us override the `TypeVariableSubstitutor`, and [we do so](https://github.com/jspecify/nullness-checker-for-checker-framework/commit/3db76159dd7537603e79dc020dffdb7495fc01cb#diff-ce82a59b06d0d82bfc6d06835d1b06ac15619bc3a71617d32189d041f02ca9c0R160).

But our larger changes to substitution rules require changes to behavior even "before" substitution occurs. Notable, `@NullnessUnspecified T` (which is how we currently represent the type of a plain `T` when it appears in non-null-aware code) does not always have a nullness of "unspecified." That's because, if the type is instantiated with `T` set to a nullable type, the result must be `@Nullable T`. Therefore, it's not guaranteed safe to, e.g., dereference a value of type `@NullnessUnspecified T`, even in "lenient mode."

To upstream our current changes, we probably need to add methods to `AnnotatedTypeFactory` for the `fixupBoundAnnotations` methods in `AnnotatedTypeMirror` and for `AsSuperVisitor` to call.

Additionally, though, note that these changes do not make sense in isolation. They may invalidate assumptions in other parts of CF, so perhaps other code should be updated, too. (For example, is any of this related to https://github.com/jspecify/nullness-checker-for-checker-framework/commit/89725d4ac9f70571339420857799f7840f9a4489#diff-ce82a59b06d0d82bfc6d06835d1b06ac15619bc3a71617d32189d041f02ca9c0R351? Or maybe that's caused by some _other_ change I've made? Or maybe it will go away once we have dataflow implemented?) Or maybe the changes are just too weird to upstream without significantly more work, which we'd have to discuss.

Links:
- see "assumed parametric nullness" in https://docs.google.com/document/d/15NND5nBxMkZ-Us6wz3Pfbt4ODIaWaJ6JDs4w6h9kUaY/edit#heading=h.u8phmghg96wr
- https://jspecify.dev/tsttcpw.html#augmented-type-of-a-type-usage-in-code
- https://jspecify.dev/tsttcpw.html#substitution

**Comment by cpovirk** [comment]
I just made another change in service of our unusual substitution rules. This one was "only" to formatting annotations, but of course that can be kind of important sometimes :)

See https://github.com/jspecify/checker-framework/commit/5fc6c2807617eeaf415d2acd69fc11ace50d98d5.

**Comment by cpovirk** [comment]
And another related change to formatting: https://github.com/jspecify/checker-framework/commit/e10a88429e022adfef2781398feae42ee0c310cc.

**Comment by cpovirk** [comment]
It might be that the _right_ way to upstream this includes changing `fixupBoundAnnotations()` to:
- modify the _lower_ bound (but leave the upper bound untouched)
- remove the annotation from the type-variable/wildcard usage itself

If we're lucky, such a change could eliminate the need for most of our existing nullness subtyping logic.

Aside: Even if we remove that logic, it's still nice to have _written_ it:

- It has helped validate our possible spec rules.
- It has helped validate our sample inputs.
- It could provide a jumping-off point for other non-GPL tools.

**Comment by cpovirk** [comment]
(At this point, though, I am just handwaving: "Modify the lower bound" might mean "replace it with the value from the type-variable wildcard usage," or it might mean something more sophisticated with glb. Also, I've been thinking mostly in terms of `? extends` wildcards (and "normal" type variables). Maybe things get more complicated in the case of `? super` wildcards and their captures?)

**Comment by cpovirk** [comment]
Also https://github.com/jspecify/checker-framework/commit/ef726d77a7f0694749acb4e9b10aa89277e0c57f (which notes that I should do better, as already discussed in the comments above).

**Comment by cpovirk** [comment]
This continues to the most concerning of our diffs from upstream. It periodically causes problems. I've made a few attempts to do things "the right way" as outlined [above](https://github.com/jspecify/checker-framework/issues/4#issuecomment-729147609), but I've never quite gotten it to work. Hopefully I will someday.

Today in workarounds:

- https://github.com/jspecify/nullness-checker-for-checker-framework/commit/ace2b513a40d7539b3c69afcd6697d5c42f506bf
- https://github.com/jspecify/nullness-checker-for-checker-framework/commit/9150a1f439e670a97f25bd1e876368d4e3fbbd09

**Comment by cpovirk** [comment]
Also: https://github.com/jspecify/nullness-checker-for-checker-framework/commit/9348e76115088149d55d9495a5c720a5f1d6cd34

**Comment by cpovirk** [comment]
Possibly also https://github.com/jspecify/nullness-checker-for-checker-framework/commit/559ba69c20d5e2f2b828032ee82a996f3a949eee, which notes that the problems there may run deeper than what I patched in that commit.

**Comment by cpovirk** [comment]
I will likely give "the right way" another attempt after CF implements capture conversion. (They have an open PR.)

I say that after having made a few unsuccessful attempts at handling one edge case:
https://github.com/jspecify/jspecify/blob/08a36665a9d5897b7f29362091a0ac8cd5970392/samples/CastOfCaptureOfUnboundedWildcardForObjectBoundedTypeParameter.java#L24

Those attempts included https://github.com/jspecify/nullness-checker-for-checker-framework/commit/9348e76115088149d55d9495a5c720a5f1d6cd34 and https://github.com/jspecify/nullness-checker-for-checker-framework/commit/364a2b1839263318f604f2e230012ae90d2477da, both of which I reverted after finding that they caused other problems. The attempts also included a couple stabs [at the wildcard handling of `CFAbstractAnalysis.createAbstractValue`](https://github.com/jspecify/checker-framework/blob/ad7d13293db0d6901aa5d14ed4568149bbeae4d9/framework/src/main/java/org/checkerframework/framework/flow/CFAbstractAnalysis.java#L149) (and, had that worked, possibly its type-variable handling, too). (I actually thought I had that working -- or at least producing a different error instead, which https://github.com/jspecify/nullness-checker-for-checker-framework/commit/559ba69c20d5e2f2b828032ee82a996f3a949eee was supposed to address? -- but I don't see it working now, so either I got confused, or I needed an additional part of the change. Hmm, maybe what I needed was my experimental implementation of `NullSpecTransfer.visitTypeCast`? That might actually have been it. Still, I've sunk way too much time into a case that I have seen only once so far in the real world -- particularly when that case might well go away if we solve the more general problem.

**Comment by cpovirk** [comment]
Note that doing this "the right way" will likely involve reworking [how we _print_ types](https://github.com/jspecify/nullness-checker-for-checker-framework/blob/2b258cb00a8e7e5884d97020b5967e918ff3872b/src/main/java/com/google/jspecify/nullness/NullSpecAnnotatedTypeFactory.java#L1118): Currently, we can always look up the "operator" (essentially, the "annotation") for a wildcard or type-variable usage directly on that type. Plus, we can assume that the bounds (which we print only in the case of a wildcard) are unmodified from the original bounds. But doing this "the right way" means removing the operator from the type itself (at least in the case of wildcards... I think?) _and_ modifying the original bounds.

I'm vaguely optimistic that we will be able to (a) look up the original bounds and (b) "diff" them against the current bounds. Together, that may let us determine the operator/annotation that we actually want to print.

**Comment by cpovirk** [comment]
I speculated yet again that we're causing ourselves issues. This time it was in https://github.com/jspecify/nullness-checker-for-checker-framework/commit/31e17113c84794ca17e55ebbec137c6839283286#diff-1a347ba7cd35402d433b3ba1b9104e9f43add180c67b64844019b8eaccbedc8fR894 (and other comments above).

**Comment by cpovirk** [comment]
Also: https://github.com/jspecify/nullness-checker-for-checker-framework/commit/0fa0a03a9ceec5aa00e2cf21819e24e4675a29b9

**Comment by cpovirk** [comment]
Probably also https://github.com/jspecify/checker-framework/commit/0a2de06cf031d89807311ea2687ef834f330d3f4
